File: static/js/grupo.js
// ═══════════════════════════════════════════════════════════════════
//  grupo.js — Visão geral do grupo ("O Grupo")
//
//  Os heróis lado a lado: vida, mana, dados de vida, condições, carga e quem
//  tem escolha de nível pendente. É a tela para decidir descanso e divisão de
//  itens sem abrir a ficha de cada um.
//
//  REGRA DE OURO, a mesma das outras telas: nenhuma regra de jogo aqui. Quem
//  o descanso curto ajuda, quem já pode dormir, a folga de carga antes de
//  ficar sobrecarregado e o resumo vêm prontos do motor (rpg/grupo.py via
//  /api/party/overview). "Pedir descanso" manda a fala do jogador ao mestre,
//  que decide se a ficção permite e abre a tela de descanso.
//
//  Não abre sozinha; aberta, a fila de telas a redesenha.
// ═══════════════════════════════════════════════════════════════════
(function () {
  'use strict';

  let _open = false;
  let _last = {};

  const esc = (s) => (window.escapeHtml ? window.escapeHtml(s) : String(s == null ? '' : s));
  const q = (id) => document.getElementById(id);
  const aspas = (s) => esc(s).replace(/'/g, "\\'");
  const maiuscula = (s) => { s = String(s || ''); return s.charAt(0).toUpperCase() + s.slice(1); };

  function ensureDom() {
    if (q('grupo-overlay')) return;
    const o = document.createElement('div');
    o.id = 'grupo-overlay';
    o.className = 'hidden';
    o.innerHTML = `
      <div id="grp-frame" role="dialog" aria-modal="true" aria-labelledby="grp-titulo">
        <header class="lcl-header">
          <button class="lcl-close" onclick="window.Grupo._fechar()" aria-label="Fechar" title="Fechar">✕</button>
          <h1 class="lcl-title" id="grp-titulo">O Grupo</h1>
          <div id="grp-hora" class="grp-hora"></div>
        </header>

        <section id="grp-resumo" class="grp-resumo" aria-label="Resumo do grupo"></section>

        <div class="grp-corpo">
          <div id="grp-cartoes" class="grp-cartoes"></div>
        </div>

        <div class="lcl-rodape">
          <div id="grp-msg" class="lcl-msg" aria-live="polite"></div>
          <button class="lcl-fechar" onclick="window.Grupo._fechar()">Fechar</button>
        </div>
      </div>`;
    o.addEventListener('click', (e) => { if (e.target === o) fechar(); });
    document.body.appendChild(o);
  }

  async function getState() {
    const f = (window.authFetch || fetch);
    const r = await f(`${window.API || ''}/api/party/overview`);
    return r.json();
  }

  // ---- Render ------------------------------------------------------
  function barra(rotulo, atual, max, pct, cls, extra, marcador) {
    return `
      <div class="grp-barra ${cls || ''}">
        <div class="grp-barra-rotulo"><span>${rotulo}</span><span>${atual}/${max}${extra || ''}</span></div>
        <div class="grp-barra-trilho">
          <div class="grp-barra-fill" style="width:${pct}%"></div>
          ${marcador != null ? `<span class="grp-barra-marca" style="left:${marcador}%"
               title="A partir daqui fica sobrecarregado"></span>` : ''}
        </div>
      </div>`;
  }

  function marcas(h) {
    const m = [];
    if (h.morto) m.push('<span class="hro-marca hro-marca-perigo">Morto</span>');
    if (h.testes_de_morte) {
      m.push(`<span class="hro-marca hro-marca-perigo">Caído: ${h.testes_de_morte.sucessos}S ${h.testes_de_morte.falhas}F</span>`);
    }
    (h.condicoes || []).forEach(c => m.push(
      `<span class="hro-marca hro-marca-perigo">${esc(maiuscula(c.nome))}${c.duracao ? ` (${c.duracao}t)` : ''}</span>`));
    if (h.exaustao) m.push(`<span class="hro-marca hro-marca-perigo">Exaustão ${h.exaustao}</span>`);
    if (h.concentracao) m.push(`<span class="hro-marca">Concentrado: ${esc(h.concentracao)}</span>`);
    (h.efeitos || []).forEach(e => m.push(`<span class="hro-marca hro-marca-efeito">${esc(e)}</span>`));
    return m.join('');
  }

  function carga(h) {
    const c = h.carga;
    const estado = c.estado === 'imovel' ? 'imóvel'
      : c.estado === 'sobrecarregado' ? 'sobrecarregado'
      : c.perto_do_limite ? `perto do limite · folga ${c.folga_kg} kg`
      : `folga ${c.folga_kg} kg`;
    const cls = c.estado !== 'livre' ? 'grp-barra-perigo' : (c.perto_do_limite ? 'grp-barra-alerta' : 'grp-barra-carga');
    return barra('Carga', c.kg, `${c.capacidade} kg`, c.pct, cls, '', c.pct_limite)
      + `<p class="grp-linha grp-carga-estado">${esc(estado)}</p>`;
  }

  function descanso(h) {
    if (h.morto) return '';
    const d = h.descanso;
    const partes = [
      d.curto_ajuda ? '<span class="grp-ok">curto ajuda</span>' : '',
      d.pode_longo ? '<span class="grp-ok">pode dormir</span>'
        : `<span class="grp-espera">longo em ${d.faltam_horas}h</span>`,
    ].filter(Boolean);
    return `<p class="grp-linha grp-descanso">${partes.join(' · ')}</p>`;
  }

  function cartao(h) {
    const v = h.vida;
    const nome = aspas(h.nome);
    const nivel = h.nivel_pendente
      ? `<button class="grp-nivel" onclick="window.Grupo._tela('${nome}','nivel')"
                 title="${h.pode_subir ? 'XP suficiente para subir de nível' : 'Há escolha de nível por fazer'}">
           ${h.pode_subir ? 'Subir de nível' : 'Escolha de nível'}</button>`
      : '';
    const vidaCls = v.atual === 0 ? 'grp-barra-perigo' : (v.pct <= 50 ? 'grp-barra-alerta' : 'grp-barra-vida');
    return `
      <article class="grp-cartao${h.morto ? ' grp-cartao-morto' : ''}${h.precisa.length ? ' grp-cartao-precisa' : ''}"
               data-nome="${esc(h.nome)}">
        <header class="grp-cartao-cabeca">
          <button class="grp-nome" onclick="window.Grupo._tela('${nome}','ficha')" title="Abrir a ficha">${esc(h.nome)}</button>
          <span class="grp-sub">${esc(maiuscula(h.classe))} · nível ${h.nivel} · CA ${h.ca}</span>
          ${nivel}
        </header>
        <div class="grp-marcas">${marcas(h)}</div>
        ${barra('Vida', v.atual, v.max, v.pct, vidaCls, v.temp ? ` <small>+${v.temp}</small>` : '')}
        ${h.mana.max ? barra('Mana', h.mana.atual, h.mana.max, h.mana.pct, 'grp-barra-mana') : ''}
        <p class="grp-linha"><span>Dados de vida</span>
          <span><b>${h.dados_de_vida.restantes}/${h.dados_de_vida.max}</b> ${esc(h.dados_de_vida.dado)}</span></p>
        ${descanso(h)}
        ${carga(h)}
        <p class="grp-linha"><span>${h.itens} ${h.itens === 1 ? 'item' : 'itens'}</span>
          <span>${h.moedas.ouro} po · ${h.moedas.prata} pp</span></p>
        ${h.precisa.length ? `<p class="grp-precisa">Precisa: ${h.precisa.map(esc).join(', ')}</p>` : ''}
        <footer class="grp-acoes">
          <button class="lcl-btn lcl-btn-sec" onclick="window.Grupo._tela('${nome}','mochila')">Mochila</button>
          ${h.conjura ? `<button class="lcl-btn lcl-btn-sec" onclick="window.Grupo._tela('${nome}','grimorio')">Grimório</button>` : ''}
        </footer>
      </article>`;
  }

  function resumo(d) {
    const r = d.resumo || {};
    const combate = !!d.em_combate;
    const nivel = (r.nivel_pendente || []).length
      ? `<p class="grp-resumo-linha"><b>Nível:</b> ${esc(r.nivel_pendente.join(', '))} com escolha pendente.</p>`
      : '';
    const semFeridos = !(r.feridos || []).length;
    const tituloCurto = combate ? 'Em combate' : (!(r.curto_ajuda || []).length ? 'Ninguém ganha com o descanso curto' : 'Manda ao mestre: Vamos fazer um descanso curto.');
    const tituloLongo = combate ? 'Em combate' : (!r.longo_disponivel ? 'Ninguém pode fazer o descanso longo ainda' : 'Manda ao mestre: Vamos acampar para um descanso longo.');
    return `
      <p class="grp-resumo-linha"><b>Descanso:</b> ${esc(r.descanso || '')}</p>
      ${r.carga ? `<p class="grp-resumo-linha"><b>Itens:</b> ${esc(r.carga)}</p>` : ''}
      ${nivel}
      <div class="grp-resumo-acoes">
        <button id="grp-curto" class="lcl-btn lcl-btn-ir" onclick="window.Grupo._descanso('curto')"
                ${combate || semFeridos || !(r.curto_ajuda || []).length ? 'disabled' : ''}
                title="${esc(tituloCurto)}">Pedir descanso curto</button>
        <button id="grp-longo" class="lcl-btn lcl-btn-ir" onclick="window.Grupo._descanso('longo')"
                ${combate || semFeridos || !r.longo_disponivel ? 'disabled' : ''}
                title="${esc(tituloLongo)}">Pedir descanso longo</button>
      </div>`;
  }

  function render(d) {
    _last = d || {};
    ensureDom();
    const herois = _last.herois || [];
    q('grp-hora').textContent = [_last.hora, _last.em_combate ? 'em combate' : ''].filter(Boolean).join(' · ');
    q('grp-resumo').innerHTML = herois.length ? resumo(_last) : '';
    q('grp-cartoes').innerHTML = herois.length
      ? herois.map(cartao).join('')
      : '<div class="lcl-vazio">Ninguém do grupo tem ficha ainda.</div>';
    mensagem('');
  }

  function mensagem(txt, erro) {
    const el = q('grp-msg');
    if (!el) return;
    el.textContent = txt || '';
    el.classList.toggle('lcl-msg-erro', !!erro);
  }

  // ---- Abrir / fechar ----------------------------------------------
  async function abrir() {
    ensureDom();
    if (!_open) {
      q('grupo-overlay').classList.remove('hidden');
      document.body.classList.add('grupo-on');
      _open = true;
    }
    mensagem('Carregando…');
    try {
      render(await getState());
    } catch (_) {
      mensagem('Falha de conexão.', true);
    }
  }

  function fechar() {
    const el = q('grupo-overlay');
    if (el) el.classList.add('hidden');
    document.body.classList.remove('grupo-on');
    _open = false;
    window.dispatchEvent(new Event('rpg:tela-fechou'));
  }

  // Aberta, redesenha com o que o mestre mudou no chat (dano, item, XP).
  async function sync() {
    if (!_open) return;
    try { render(await getState()); } catch (_) { /* nunca derruba o turno */ }
  }

  function tela(nome, qual) {
    fechar();
    if (qual === 'ficha' && window.Herois) window.Herois._abrir(nome);
    if (qual === 'mochila' && window.Inventory) window.Inventory._abrir(nome);
    if (qual === 'grimorio' && window.Grimoire) window.Grimoire._abrir(nome);
    if (qual === 'nivel' && window.LevelUp) { window.LevelUp._trocar(nome); window.LevelUp._abrir(); }
  }

  // A fala vai ao mestre como se o jogador a tivesse digitado; ele decide se
  // a ficção permite e abre a tela de descanso com offer_rest.
  function pedirDescanso(tipo) {
    if (typeof waiting !== 'undefined' && waiting) {
      mensagem('Aguarde o mestre terminar de responder.', true);
      return;
    }
    if (typeof window.sendToAgent !== 'function' || typeof window.appendUser !== 'function') return;
    const texto = tipo === 'longo'
      ? 'Vamos acampar para um descanso longo.'
      : 'Vamos fazer um descanso curto.';
    fechar();
    window.appendUser(texto);
    window.sendToAgent(texto, true);
  }

  window.Grupo = {
    sync,
    _abrir: abrir,
    _fechar: fechar,
    _tela: tela,
    _descanso: pedirDescanso,
    _estado: () => _last,
  };

  document.addEventListener('keydown', (e) => { if (e.key === 'Escape' && _open) fechar(); });
  document.addEventListener('DOMContentLoaded', () => { ensureDom(); });
})();

File: rpg/grupo.py
"""
grupo.py
Visão geral do grupo: os heróis lado a lado.

Para decidir DESCANSO e DIVISÃO DE ITENS o jogador abria a ficha de cada um,
uma por vez, e fazia a conta de cabeça: quem está ferido, quem ainda tem dado
de vida, quem já pode dormir, quem aguenta carregar a cota de malha.

Nenhuma regra nova aqui. Cada número vem da mesma conta do resto do jogo:
a ficha do herói (hero_snapshot), a tela de descanso (reserva de dados, as 24
horas do descanso longo, o que a noite devolve) e a carga (capacidade e
metade dela, a partir da qual o herói fica sobrecarregado). O que este módulo
acrescenta é o JUNTAR: o que cada um precisa e o resumo do grupo.
"""

from __future__ import annotations

from rpg import memory


class _MotorTardio:
    """tools_dnd importa muita coisa; carregar só quando a tela pede."""

    def __getattr__(self, nome):
        from rpg import tools_dnd
        return getattr(tools_dnd, nome)


td = _MotorTardio()

# Carga a partir da qual o cartão avisa "perto do limite": 80% do ponto em que
# o herói fica sobrecarregado (metade da capacidade).
_PERTO_DO_LIMITE = 0.8


def _pct(atual: int, maximo: int) -> int:
    return max(0, min(100, round(atual / maximo * 100))) if maximo > 0 else 0


def _carga(ch: dict) -> dict:
    estado, kg, cap = td._estado_de_carga(ch)
    limite = round(cap / 2, 1)
    return {
        "kg": round(kg, 1),
        "capacidade": cap,
        "limite_sobrecarga": limite,
        "folga_kg": round(max(0.0, limite - kg), 1),
        "estado": estado,
        "perto_do_limite": estado == "livre" and limite > 0 and kg >= limite * _PERTO_DO_LIMITE,
        "pct": _pct(kg, cap),
        "pct_limite": 50,
    }


def _heroi(ch: dict) -> dict:
    nome = ch.get("name", "")
    p = td.hero_snapshot(nome)["personagem"]
    s = ch["sheet"]
    morto = (ch.get("status") or "").lower() == "morto"
    pode_longo, faltam = td._situacao_descanso_longo(s)
    vida, mana, dados = p["vida"], p["mana"], p["dados_de_vida"]

    # O que falta a ele, na ordem em que pesa numa decisão de descanso.
    precisa = []
    if not morto:
        if vida["atual"] == 0:
            precisa.append("caído")
        elif vida["atual"] < vida["teto"]:
            precisa.append(f"vida {vida['atual']}/{vida['max']}")
        if mana["max"] and mana["atual"] < mana["max"]:
            precisa.append(f"mana {mana['atual']}/{mana['max']}")
        if p["exaustao"]:
            precisa.append(f"exaustão {p['exaustao']}")

    return {
        "nome": nome,
        "classe": p["classe"],
        "raca": p["raca"],
        "nivel": p["nivel"],
        "morto": morto,
        "vida": {**vida, "pct": _pct(vida["atual"], vida["max"])},
        "mana": {**mana, "pct": _pct(mana["atual"], mana["max"])},
        "ca": p["ca"],
        "dados_de_vida": {**dados, "no_longo": td._dados_devolvidos_no_longo(s)},
        "exaustao": p["exaustao"],
        "testes_de_morte": p["testes_de_morte"],
        "concentracao": p["concentracao"],
        "condicoes": p["condicoes"],
        "efeitos": p["efeitos"],
        "carga": _carga(ch),
        "moedas": p["moedas"],
        "itens": p["itens"],
        "nivel_pendente": bool(p["pode_subir"] or p["escolhas_pendentes"]),
        "pode_subir": p["pode_subir"],
        "escolhas_pendentes": p["escolhas_pendentes"],
        "conjura": p["conjura"],
        "descanso": {
            "pode_longo": pode_longo and not morto,
            "faltam_horas": faltam,
            # Curto só adianta a quem está ferido e ainda tem dado para gastar.
            "curto_ajuda": (not morto and 0 < vida["atual"] < vida["teto"]
                            and dados["restantes"] > 0),
        },
        "precisa": precisa,
    }


def _nomes(lista: list[str]) -> str:
    if len(lista) <= 1:
        return "".join(lista)
    return ", ".join(lista[:-1]) + " e " + lista[-1]


def _resumo(herois: list[dict], em_combate: bool) -> dict:
    vivos = [h for h in herois if not h["morto"]]
    feridos = [h["nome"] for h in vivos if h["precisa"]]
    curto = [h["nome"] for h in vivos if h["descanso"]["curto_ajuda"]]
    sem_dado = [h["nome"] for h in vivos
                if 0 < h["vida"]["atual"] < h["vida"]["teto"] and h["dados_de_vida"]["restantes"] == 0]
    podem_longo = [h["nome"] for h in vivos if h["descanso"]["pode_longo"]]
    esperam = sorted(((h["nome"], h["descanso"]["faltam_horas"]) for h in vivos
                      if not h["descanso"]["pode_longo"]), key=lambda x: -x[1])

    if em_combate:
        descanso = "Em combate: descanso só depois da luta."
    elif not feridos:
        descanso = "Ninguém precisa de descanso agora."
    else:
        partes = []
        if curto:
            partes.append(f"O descanso curto ajuda {_nomes(curto)}.")
        if sem_dado:
            partes.append(f"{_nomes(sem_dado)} sem dado de vida: só o longo cura.")
        if podem_longo and not esperam:
            partes.append("O descanso longo está disponível para todos.")
        elif podem_longo:
            nome, horas = esperam[0]
            partes.append(f"Descanso longo: {_nomes(podem_longo)} já podem; "
                          f"{nome} só daqui a {horas}h.")
        else:
            horas = esperam[0][1] if esperam else 0
            partes.append(f"Descanso longo só daqui a {horas}h.")
        descanso = " ".join(partes)

    # Divisão de itens: quem tem mais folga antes de ficar sobrecarregado.
    por_folga = sorted(vivos, key=lambda h: -h["carga"]["folga_kg"])
    pesados = [h["nome"] for h in vivos if h["carga"]["estado"] != "livre"]
    perto = [h["nome"] for h in vivos if h["carga"]["perto_do_limite"]]
    partes = []
    if por_folga:
        partes.append("Mais folga para carregar: "
                      + ", ".join(f"{h['nome']} ({h['carga']['folga_kg']:g} kg)" for h in por_folga[:3])
                      + ".")
    if pesados:
        partes.append(f"Sobrecarregado: {_nomes(pesados)}.")
    if perto:
        partes.append(f"Perto do limite: {_nomes(perto)}.")

    return {
        "descanso": descanso,
        "carga": " ".join(partes),
        "nivel_pendente": [h["nome"] for h in vivos if h["nivel_pendente"]],
        "feridos": feridos,
        "curto_ajuda": curto,
        "longo_disponivel": bool(podem_longo),
        "mais_folga": por_folga[0]["nome"] if por_folga else "",
    }


def _sem_ficha() -> list[dict]:
    """
    Quem está no grupo sem ficha de regras: o companheiro que o mestre
    recrutou pela narrativa e nunca recebeu atributos.

    Numa campanha D&D essa gente sumia da barra lateral e da visão geral, que
    só conheciam quem tem ficha — como se não estivesse no grupo. Aqui eles
    voltam com o que existe deles: nome, papel e a descrição.
    """
    chars = memory.campaign.get("characters", {}) or {}
    papeis = {memory.char_key(m.get("name", "")): (m.get("role") or "")
              for m in (memory.campaign.get("party") or []) if isinstance(m, dict)}
    fora = []
    for ch in chars.values():
        if not memory.is_party_member(ch) or (ch.get("sheet") or {}):
            continue
        nome = ch.get("name", "")
        fora.append({
            "nome": nome,
            "papel": papeis.get(memory.char_key(nome), "") or (ch.get("role") or ""),
            "descricao": (ch.get("description") or "").strip(),
            "morto": (ch.get("status") or "").lower() == "morto",
        })
    return sorted(fora, key=lambda x: x["nome"].lower())


def group_snapshot() -> dict:
    """O grupo lado a lado e o resumo (JSON-serializável)."""
    herois = [_heroi(ch) for ch in td._grupo_com_ficha()]
    em_combate = td._em_combate()
    return {
        "herois": herois,
        "sem_ficha": _sem_ficha(),
        "em_combate": em_combate,
        "hora": td._hora_legivel(),
        "resumo": _resumo(herois, em_combate),
    }
